improver/utilities/temporal_interpolation.py

```diff
@@ -351,8 +412,107 @@ def daynight_interpolate(interpolated_cube: Cube) -> CubeList:
         daynightplugin = DayNightMask()
         daynight_mask = daynightplugin(interpolated_cube)
         index = daynight_mask.data == daynightplugin.night
-        interpolated_cube.data[..., index] = 0.0
-        return iris.cube.CubeList(list(interpolated_cube.slices_over("time")))
+
+        # Reshape the time, y, x mask to match the input which may include addtional
+        # dimensions, such as realization.
+        dropped_crds = [
+            crd
+            for crd in interpolated_cube.coords(dim_coords=True)
+            if crd not in daynight_mask.coords(dim_coords=True)
+        ]
+        if dropped_crds:
+            cslices = interpolated_cube.slices_over(dropped_crds)
+            masked_data = CubeList()
+            for cslice in cslices:
+                cslice.data[index] = 0.0
+                masked_data.append(cslice)
+            interpolated_cube = masked_data.merge_cube()
+        else:
+            interpolated_cube.data[index] = 0.0
+
+        return CubeList(list(interpolated_cube.slices_over("time")))
```

Comment:
Is there a reason to not broadcast `index` to the same shape as `interpolated_cube` and then call `interpolated_cube.data[index] = 0.0` once, rather than looping over the non-common dim_coords?

Reply:
Generally slicing helps us avoid swelling the data in memory. Working with ECMWF data with 51 realizations I am constantly scared of this. It may well be that the merge_cube step requires everything in memory anyway, but where I can reasonably slice, I slice :-)

Reply:
Fair enough, that's a perfectly sensible justification.